File: src/backend/utils/logger.py
import logging
import sys
import os
from typing import Any

def truncate_llm_response(response: Any, head_tail_length: int = 100) -> str:
    """Truncate LLM response for debug logging, showing both start and end of the response"""
    if not response:
        return str(response)
        
    # Convert response to string if it's not already
    response_str = str(response)
        
    if len(response_str) <= head_tail_length * 2:
        return response_str
        
    head = response_str[:head_tail_length]
    tail = response_str[-head_tail_length:]
    middle_length = len(response_str) - (head_tail_length * 2)
    
    return f"{head}... [{middle_length} chars truncated] ...{tail}"

def setup_logger(name: str) -> logging.Logger:
    """Setup logger with console handler"""
    logger = logging.getLogger(name)
    
    # Set log level based on DEBUG_MODE environment variable
    debug_mode = os.getenv('DEBUG_MODE', 'false').lower() == 'true'
    logger.setLevel(logging.DEBUG if debug_mode else logging.INFO)

    # Format for console output
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )

    # Console Handler
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)

    return logger

# Create main logger instance
logger = setup_logger('army-gpt') 

File: src/backend/utils/logger.ts
import winston from 'winston';
import { config } from '../config.js';

// Types
type LogLevel = 'debug' | 'info' | 'warn' | 'error';

// Helper function to truncate long responses
export function truncateLLMResponse(response: any, headTailLength = 100): string {
  if (!response) {
    return String(response);
  }
  
  const responseStr = String(response);
  
  if (responseStr.length <= headTailLength * 2) {
    return responseStr;
  }
  
  const head = responseStr.slice(0, headTailLength);
  const tail = responseStr.slice(-headTailLength);
  const middleLength = responseStr.length - (headTailLength * 2);
  
  return `${head}... [${middleLength} chars truncated] ...${tail}`;
}

// Create logger instance
const loggerInstance = winston.createLogger({
  level: config.server.logLevel,
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.printf(({ timestamp, level, message, ...meta }) => {
      const metaStr = Object.keys(meta).length ? JSON.stringify(meta) : '';
      return `${timestamp} - army-gpt - ${level}: ${message} ${metaStr}`;
    })
  ),
  transports: [
    new winston.transports.Console({
      format: winston.format.combine(
        winston.format.colorize(),
        winston.format.simple()
      )
    })
  ]
});

// Add debug mode warning if enabled
if (config.server.debug) {
  loggerInstance.warn('Debug mode is enabled - verbose logging will be shown');
}

// Extend logger with isLevelEnabled method
interface ExtendedLogger extends winston.Logger {
  isLevelEnabled(level: LogLevel): boolean;
}

const logger = loggerInstance as ExtendedLogger;

// Add isLevelEnabled method
logger.isLevelEnabled = function(level: LogLevel): boolean {
  const currentLevel = this.level;
  const levels = this.levels;
  return levels[currentLevel] >= levels[level];
};

export { logger }; 